dataland-api-key-manager/src/main/kotlin/org/dataland/datalandapikeymanager/services/ApiKeyManager.kt

```diff
@@ -0,0 +1,231 @@
+        return hash
+    }
+
+    private fun generateRandomByteArray(length: Int): ByteArray {
+        val bytes = ByteArray(length)
+        SecureRandom().nextBytes(bytes)
+        return bytes
+    }
+
+    private fun generateSalt(): ByteArray {
+        return generateRandomByteArray(saltByteLength)
+    }
+
+    private fun generateApiKeySecretAndEncodeToHex(): String {
+        return HexFormat.of().formatHex(generateRandomByteArray(keyByteLength))
+    }
+
+    private fun getAuthentication(): Authentication {
+        return SecurityContextHolder.getContext().authentication
+    }
+
+    private fun getKeycloakUserIdBase64Encoded(authentication: Authentication): String {
+        return encodeToBase64(authentication.name.toByteArray(utf8Charset))
+    }
+
+    private fun calculateExpiryDate(daysValid: Int?): Long? {
+        return if (daysValid == null)
+            null else
+            Instant.now().epochSecond
+    }
+
+    private fun generateApiKeyMetaInfo(daysValid: Int?): ApiKeyMetaInfo {
+        val authentication = getAuthentication()
+        val keycloakUserId = authentication.name
+        val keycloakRoles = authentication.authorities.map { it.authority!! }.toList()
+        return ApiKeyMetaInfo(keycloakUserId, keycloakRoles, calculateExpiryDate(daysValid))
+    }
+
+    /**
+     * A method that generates an API key which is valid for the specified number of days
+     * @param daysValid the number of days the API key should be valid from time of generation
+     * @return the API key and its meta info
+     */
+    fun generateNewApiKey(daysValid: Int?): ApiKeyAndMetaInfo {
+        val apiKeyMetaInfo = generateApiKeyMetaInfo(daysValid)
+
+        val newSalt = generateSalt()
+        val keycloakUserIdBase64Encoded = encodeToBase64(apiKeyMetaInfo.keycloakUserId!!.toByteArray(Charsets.UTF_8))
+        val newApiKeyWithoutCrc32Value = keycloakUserIdBase64Encoded + "_" + generateApiKeySecretAndEncodeToHex()
+        val newCrc32Value = EncodingUtils.calculateCrc32Value(newApiKeyWithoutCrc32Value.toByteArray(utf8Charset))
+        val newApiKey = newApiKeyWithoutCrc32Value + "_" + newCrc32Value
+        val newApiKeyHashedAndBase64Encoded = encodeToBase64(hashString(newApiKey, newSalt))
+
+        val hashedAndBase64EncodedApiKeyEntityToStore = StoredHashedAndBase64EncodedApiKeyEntity(
+            newApiKeyHashedAndBase64Encoded,
+            keycloakUserIdBase64Encoded,
+            apiKeyMetaInfo.keycloakRoles!!,
+            apiKeyMetaInfo.expiryDate,
+            encodeToBase64(newSalt)
+        )
+
+        storedHashedAndBase64EncodedApiKeyRepository.save(hashedAndBase64EncodedApiKeyEntityToStore)
+        logger.info(
+            "Generated Api Key with hashed value $newApiKeyHashedAndBase64Encoded and meta info $apiKeyMetaInfo."
+        )
+        return ApiKeyAndMetaInfo(newApiKey, apiKeyMetaInfo)
+    }
+
+    private fun checkIfApiKeyForUserIsCorrectAndReturnApiKeyMetaInfoWithActivityStatus(
```

Comment:
Den Namen finde ich nicht sehr gelungen. Zudem sollten Funktionen immer genau eine Sache tun nicht mehrere.

Reply:
Ok ich schaue nochmal, wie ich das sauberer machen kann.

Reply:
Man könnte die Besitmmung des acitvitStatus hier rausziehen und in eine eigene FUnktion packen.

Der Rest gehört meines Erachtes hier rein. Dieser besteht ja nur aus 
"checke ob Api Key übereinstimmt => wenn nein, schicke dies zurück, wenn ja, schicke jenes zurück"

Reply:
Ist obsolet